Opening ics file fails

Hi!
I have a setup running magic mirror 2.3.1.0 with node 23.11.0 and found an issue in its calendar plugin when opening certain ics files. Since the calendar plugin uses the node-ical module I'm writing my issue here.
The symptom looks as follows:
`[ERROR] Calendar Error. Could not fetch calendar:  http://localhost:8080/CALDAV/HOME_home.ics TypeError: duration is not iterable`
`at Object.originalEnd (/home/admin/MagicMirror/node_modules/node-ical/ical.js:481:29)`
`at Object.END (/home/admin/MagicMirror/node_modules/node-ical/ical.js:652:26)`
`at Object.handleObject (/home/admin/MagicMirror/node_modules/node-ical/ical.js:696:39)`
`at Object.parseLines (/home/admin/MagicMirror/node_modules/node-ical/ical.js:748:18)`
`at Object.parseICS (/home/admin/MagicMirror/node_modules/node-ical/ical.js:784:18)`
`at sync.parseICS (/home/admin/MagicMirror/node_modules/node-ical/node-ical.js:198:15)`
`at autodetect.parseICS (/home/admin/MagicMirror/node_modules/node-ical/node-ical.js:229:17)`
`at /home/admin/MagicMirror/modules/default/calendar/calendarfetcher.js:58:18`
`at process.processTicksAndRejections (node:internal/process/task_queues:105:5)`

The issue arises when opening a file that has no specified duration length like this:

BEGIN:VEVENT
DTSTAMP:20250525T092115Z
UID:a431dc0c-19a5-41a5-96ad-3a07bc58cf81
SUMMARY:Travels
STATUS:CONFIRMED
DTSTART:20250706T160000Z
DURATION:P
END:VEVENT

The duration is set to P but nothing else. Since the duration is not undefined in the code, it tries to check for the length of the duration creating a list from a regex match. This list is emty, because of the zero time duration and throws the error when it wants to iterate over the list.

My current fix was in line 481 from the ical.js to switch form
`if (curr.duration !== undefined) {`
to
`if (curr.duration !== undefined && curr.duration !== 'P') {`
This checks if a duration is specified and if not treats it as undefined. I did not test all possible options though...
That way it skips the process of determining the end date of an event and uses the start date as end date.
It's a rather quick fix, but it does the trick I would say. Maybe there is a better option/location to handle events with no specified duration that are not undefined. I didn't check in depth if the `curr.duration` could be set to `undefined` when the file is loaded, but this works for now.
If further info is neede, let me know!
Best

--- a/ical.js
+++ b/ical.js
@@ -519,45 +519,67 @@ module.exports = {
         const par = stack.pop();
 
         if (!curr.end) { // RFC5545, 3.6.1
-          // Set the end according to the datetype of event
-          curr.end = (curr.datetype === 'date-time') ? new Date(curr.start) : tzUtil.utcAdd(curr.start, 1, 'days');
+          // Helper: clone a Date and preserve custom metadata (tz, dateOnly)
+          const cloneDateWithMeta = (source, newTime = source) => {
+            const cloned = new Date(newTime);
+            if (source?.tz) {
+              cloned.tz = source.tz;
+            }
 
-          // If there was a duration specified
-          // see RFC5545, 3.3.6 (no year and month)
-          if (curr.duration !== undefined) {
-            const durationUnits
-            = {
-              W: 'weeks',
-              D: 'days',
-              H: 'hours',
-              M: 'minutes',
-              S: 'seconds',
-            };
-            // Get the list of duration elements
-            const duration = curr.duration.match(/-?\d{1,10}[WDHMS]/g);
-            if (!duration || duration.length === 0) {
-              // Invalid or empty DURATION: treat as zero duration (end = start)
-              // This handles malformed cases like "DURATION:P" gracefully (Postel's Law)
-              console.warn(`[node-ical] Ignoring malformed DURATION value: "${curr.duration}" – treating as zero duration`);
-              curr.end = curr.start;
-            } else {
-              // Use the duration to create the end value, from the start
-              let newEnd = curr.start;
+            if (source?.dateOnly) {
+              cloned.dateOnly = source.dateOnly;
+            }
 
-              // Is the 1st character a negative sign?
-              const indicator = curr.duration.startsWith('-') ? -1 : 1;
+            return cloned;
+          };
 
-              for (const r of duration) {
-                const unit = r.slice(-1);
-                if (!durationUnits[unit]) {
-                  throw new Error(`Invalid duration unit: ${unit}`);
-                }
+          // Helper: extract string value from DURATION (handles {params, val} shape)
+          const getDurationString = duration => {
+            if (typeof duration === 'object' && duration?.val) {
+              return String(duration.val);
+            }
 
-                newEnd = tzUtil.utcAdd(newEnd, Number.parseInt(r, 10) * indicator, durationUnits[r.toString().slice(-1)]);
+            if (duration) {
+              return String(duration);
+            }
+
+            return '';
+          };
+
+          // Calculate end date based on DURATION or default rules
+          if (curr.duration === undefined) {
+            // No DURATION: default end is same time (date-time) or +1 day (date-only)
+            curr.end = curr.datetype === 'date-time'
+              ? cloneDateWithMeta(curr.start)
+              : cloneDateWithMeta(curr.start, tzUtil.utcAdd(curr.start, 1, 'days'));
+          } else {
+            const durationString = getDurationString(curr.duration);
+            const durationParts = durationString.match(/-?\d{1,10}[WDHMS]/g);
+
+            if (durationParts && durationParts.length > 0) {
+              // Valid DURATION: apply each component (W/D/H/M/S)
+              const units = {
+                W: 'weeks',
+                D: 'days',
+                H: 'hours',
+                M: 'minutes',
+                S: 'seconds',
+              };
+              const sign = durationString.startsWith('-') ? -1 : 1;
+
+              let endTime = curr.start;
+              for (const part of durationParts) {
+                const value = Number.parseInt(part, 10) * sign;
+                const unit = units[part.slice(-1)];
+                endTime = tzUtil.utcAdd(endTime, value, unit);
               }
 
-              // End is a Date type, not moment
-              curr.end = new Date(newEnd);
+              curr.end = cloneDateWithMeta(curr.start, endTime);
+            } else {
+              // Malformed DURATION (e.g., "P", "PT", "") → treat as zero duration
+              // Follows Postel's Law: be liberal in what you accept
+              console.warn(`[node-ical] Ignoring malformed DURATION value: "${durationString}" – treating as zero duration`);
+              curr.end = cloneDateWithMeta(curr.start);
             }
           }
         }

--- a/ical.js
+++ b/ical.js
@@ -536,26 +536,29 @@ module.exports = {
             // Get the list of duration elements
             const duration = curr.duration.match(/-?\d{1,10}[WDHMS]/g);
             if (!duration || duration.length === 0) {
-              throw new Error('Invalid DURATION format: ' + curr.duration);
-            }
+              // Invalid or empty DURATION: treat as zero duration (end = start)
+              // This handles malformed cases like "DURATION:P" gracefully (Postel's Law)
+              console.warn(`[node-ical] Ignoring malformed DURATION value: "${curr.duration}" – treating as zero duration`);
+              curr.end = curr.start;
+            } else {
+              // Use the duration to create the end value, from the start
+              let newEnd = curr.start;
 
-            // Use the duration to create the end value, from the start
-            let newEnd = curr.start;
+              // Is the 1st character a negative sign?
+              const indicator = curr.duration.startsWith('-') ? -1 : 1;
 
-            // Is the 1st character a negative sign?
-            const indicator = curr.duration.startsWith('-') ? -1 : 1;
+              for (const r of duration) {
+                const unit = r.slice(-1);
+                if (!durationUnits[unit]) {
+                  throw new Error(`Invalid duration unit: ${unit}`);
+                }
 
-            for (const r of duration) {
-              const unit = r.slice(-1);
-              if (!durationUnits[unit]) {
-                throw new Error(`Invalid duration unit: ${unit}`);
+                newEnd = tzUtil.utcAdd(newEnd, Number.parseInt(r, 10) * indicator, durationUnits[r.toString().slice(-1)]);
               }
 
-              newEnd = tzUtil.utcAdd(newEnd, Number.parseInt(r, 10) * indicator, durationUnits[r.toString().slice(-1)]);
+              // End is a Date type, not moment
+              curr.end = new Date(newEnd);
             }
-
-            // End is a Date type, not moment
-            curr.end = new Date(newEnd);
           }
         }
 

--- a/test/advanced.test.js
+++ b/test/advanced.test.js
@@ -436,6 +436,30 @@ END:VCALENDAR`;
       assert.equal(event.end.toDateString(), new Date(2024, 1, 22).toDateString());
     });
 
+    // Test parameterized DURATION (shape-safety)
+    it('handles parameterized DURATION values', () => {
+      const ics = `BEGIN:VCALENDAR
+VERSION:2.0
+BEGIN:VEVENT
+UID:parameterized-duration-test
+DTSTART:20250101T120000Z
+DURATION;X-CUSTOM=foo:P1D
+SUMMARY:Test Parameterized Duration
+END:VEVENT
+END:VCALENDAR`;
+
+      const data = ical.parseICS(ics);
+      const event = Object.values(data).find(x => x.type === 'VEVENT');
+
+      assert.ok(event, 'Event should be parsed');
+      assert.ok(event.start instanceof Date, 'Start should be a Date');
+      assert.ok(event.end instanceof Date, 'End should be a Date');
+
+      // Should add 1 day despite having parameters
+      const expectedEnd = new Date(event.start.getTime() + (24 * 60 * 60 * 1000));
+      assert.equal(event.end.toISOString(), expectedEnd.toISOString());
+    });
+
     // Issue #381 – empty DURATION:P should not crash
     it('handles empty duration DURATION:P gracefully (issue #381)', () => {
       const ics = `BEGIN:VCALENDAR

--- a/test/advanced.test.js
+++ b/test/advanced.test.js
@@ -455,6 +455,10 @@ END:VCALENDAR`;
       // Instead: treat empty/invalid duration as zero duration (end = start)
       // This follows Postel's Law: "be liberal in what you accept"
 
+      // Stub console.warn to keep test output clean
+      const originalWarn = console.warn;
+      console.warn = () => {};
+
       const data = ical.parseICS(ics);
       const event = Object.values(data).find(x => x.type === 'VEVENT');
 
@@ -469,10 +473,17 @@ END:VCALENDAR`;
         event.start.toISOString(),
         'End should equal start for invalid DURATION:P (zero duration)',
       );
+
+      // Restore console.warn
+      console.warn = originalWarn;
     });
 
     // Additional test: Multiple invalid duration formats should be handled
     it('handles various malformed DURATION values gracefully', () => {
+      // Stub console.warn to keep test output clean
+      const originalWarn = console.warn;
+      console.warn = () => {};
+
       const testCases = [
         // Malformed / invalid durations → should be treated as zero duration
         {duration: 'P', summary: 'Empty P', expected: 'zero'},
@@ -488,11 +499,11 @@ END:VCALENDAR`;
         {duration: 'P1DT2H', summary: 'One day two hours', expected: '1d2h'}, // Day + hours
       ];
 
-      for (const testCase of testCases) {
+      for (const [i, testCase] of testCases.entries()) {
         const ics = `BEGIN:VCALENDAR
 VERSION:2.0
 BEGIN:VEVENT
-UID:malformed-${testCase.summary}
+UID:malformed-${i}
 DTSTART:20250706T160000Z
 DURATION:${testCase.duration}
 SUMMARY:${testCase.summary}
@@ -567,6 +578,9 @@ END:VCALENDAR`;
           }
         }
       }
+
+      // Restore console.warn
+      console.warn = originalWarn;
     });
   });
 

--- a/test/advanced.test.js
+++ b/test/advanced.test.js
@@ -435,6 +435,139 @@ END:VCALENDAR`;
       assert.equal(event.start.toDateString(), new Date(2024, 1, 15).toDateString());
       assert.equal(event.end.toDateString(), new Date(2024, 1, 22).toDateString());
     });
+
+    // Issue #381 – empty DURATION:P should not crash
+    it('handles empty duration DURATION:P gracefully (issue #381)', () => {
+      const ics = `BEGIN:VCALENDAR
+VERSION:2.0
+PRODID:-//Test//Test//EN
+BEGIN:VEVENT
+UID:empty-duration-test
+DTSTART:20250706T160000Z
+DURATION:P
+SUMMARY:Travels
+STATUS:CONFIRMED
+DTSTAMP:20250525T092115Z
+END:VEVENT
+END:VCALENDAR`;
+
+      // Ideal behavior: Parser should NOT crash or throw for malformed DURATION
+      // Instead: treat empty/invalid duration as zero duration (end = start)
+      // This follows Postel's Law: "be liberal in what you accept"
+
+      const data = ical.parseICS(ics);
+      const event = Object.values(data).find(x => x.type === 'VEVENT');
+
+      assert.ok(event, 'Event should be parsed despite invalid DURATION:P');
+      assert.equal(event.summary, 'Travels');
+      assert.ok(event.start instanceof Date, 'Start should be a Date');
+      assert.ok(event.end instanceof Date, 'End should be a Date');
+
+      // Invalid/empty duration should be treated as zero duration
+      assert.equal(
+        event.end.toISOString(),
+        event.start.toISOString(),
+        'End should equal start for invalid DURATION:P (zero duration)',
+      );
+    });
+
+    // Additional test: Multiple invalid duration formats should be handled
+    it('handles various malformed DURATION values gracefully', () => {
+      const testCases = [
+        // Malformed / invalid durations → should be treated as zero duration
+        {duration: 'P', summary: 'Empty P', expected: 'zero'},
+        {duration: 'PT', summary: 'Empty PT', expected: 'zero'},
+        {duration: 'INVALID', summary: 'Invalid text', expected: 'zero'},
+        {duration: '', summary: 'Empty string', expected: 'zero'},
+        // Valid durations
+        {duration: 'PT0S', summary: 'Zero duration', expected: 'zero'}, // Valid zero
+        {duration: 'P0D', summary: 'Zero days', expected: 'zero'}, // Valid zero (days)
+        {duration: 'P1D', summary: 'One day', expected: 'oneday'}, // Valid one day
+        {duration: '-P1D', summary: 'Negative one day', expected: 'negoneday'}, // Negative duration
+        {duration: 'PT1H30M', summary: 'One hour thirty minutes', expected: '1h30m'}, // Combined
+        {duration: 'P1DT2H', summary: 'One day two hours', expected: '1d2h'}, // Day + hours
+      ];
+
+      for (const testCase of testCases) {
+        const ics = `BEGIN:VCALENDAR
+VERSION:2.0
+BEGIN:VEVENT
+UID:malformed-${testCase.summary}
+DTSTART:20250706T160000Z
+DURATION:${testCase.duration}
+SUMMARY:${testCase.summary}
+END:VEVENT
+END:VCALENDAR`;
+
+        // All durations should parse without throwing
+        const data = ical.parseICS(ics);
+        const event = Object.values(data).find(x => x.type === 'VEVENT');
+
+        assert.ok(event, `Event should be parsed for DURATION:${testCase.duration}`);
+        assert.ok(event.start instanceof Date, 'Start should be a Date');
+        assert.ok(event.end instanceof Date, 'End should be a Date');
+
+        switch (testCase.expected) {
+          case 'zero': {
+            // Invalid or zero durations: end should equal start
+            assert.equal(
+              event.end.toISOString(),
+              event.start.toISOString(),
+              `DURATION:${testCase.duration} should result in zero duration (end = start)`,
+            );
+            break;
+          }
+
+          case 'oneday': {
+            // P1D should add 1 day
+            const expectedEnd = new Date(event.start.getTime() + (24 * 60 * 60 * 1000));
+            assert.equal(
+              event.end.toISOString(),
+              expectedEnd.toISOString(),
+              `DURATION:${testCase.duration} should add 1 day`,
+            );
+            break;
+          }
+
+          case 'negoneday': {
+            // -P1D should subtract 1 day
+            const expectedEnd = new Date(event.start.getTime() - (24 * 60 * 60 * 1000));
+            assert.equal(
+              event.end.toISOString(),
+              expectedEnd.toISOString(),
+              `DURATION:${testCase.duration} should subtract 1 day`,
+            );
+            break;
+          }
+
+          case '1h30m': {
+            // PT1H30M should add 1 hour 30 minutes
+            const expectedEnd = new Date(event.start.getTime() + (90 * 60 * 1000));
+            assert.equal(
+              event.end.toISOString(),
+              expectedEnd.toISOString(),
+              `DURATION:${testCase.duration} should add 1h30m`,
+            );
+            break;
+          }
+
+          case '1d2h': {
+            // P1DT2H should add 1 day + 2 hours
+            const expectedEnd = new Date(event.start.getTime() + (26 * 60 * 60 * 1000));
+            assert.equal(
+              event.end.toISOString(),
+              expectedEnd.toISOString(),
+              `DURATION:${testCase.duration} should add 1 day + 2 hours`,
+            );
+            break;
+          }
+
+          default: {
+            assert.fail(`Unexpected test case: ${testCase.expected}`);
+          }
+        }
+      }
+    });
   });
 
   // Test_with_int_tzid.ics – integer-like tzid preserved

